Fix raw provider keys

mattcaswell opened the pull request:
The functions EVP_PKEY_get_raw_private_key and EVP_PKEY_get_raw_public_key fail with a NULL pointer deref if you pass a provider key to them. This also highlighted that EVP_PKEY_new_raw_private_key and EVP_PKEY_new_raw_public_key do not create provider side keys as you might expect.

This PR fixes both of those issues. Along the way a number of other issues were highlighted.

This is WIP because the tests are failing. At least one of those failures (possibly all of them) are because EVP_PKEY_cmp is not working with ECX provider side keys.

Fixes #11627

levitte commented on crypto/evp/p_lib.c:
```diff
@@ -317,78 +318,234 @@ int EVP_PKEY_cmp(const EVP_PKEY *a, const EVP_PKEY *b)
+            if (params == NULL) {
+                EVPerr(0, ERR_R_MALLOC_FAILURE);
+                goto err;
+            }
```
Er, I fail to see why you go through all the trouble with OSSL_PARAM_BUILD here...

``` C
        OSSL_PARAM params[] = { OSSL_PARAM_END, OSSL_PARAM_END };

        params[0] = OSSL_PARAM_construct_octet_string(priv ? OSSL_PKEY_PARAM_PRIV_KEY
                                                           : OSSL_PKEY_PARAM_PUB_KEY,
                                                      key, len);
```

Done.

mattcaswell replied:
Yes - that's much better. Done.

levitte commented on crypto/evp/p_lib.c:
```diff
@@ -317,78 +318,234 @@ int EVP_PKEY_cmp(const EVP_PKEY *a, const EVP_PKEY *b)
-    if (!ret->ameth->set_priv_key(ret, priv, len)) {
-        EVPerr(EVP_F_EVP_PKEY_NEW_RAW_PRIVATE_KEY, EVP_R_KEY_SETUP_FAILED);
+    if (pkey->ameth == NULL) {
+        /* Shouldn't happen */
```
This could be an assertion, then, i.e. an extra check that `pkey_set_type()` did its job.

mattcaswell replied:
Done

levitte commented on util/libcrypto.num:
```diff
@@ -5079,3 +5079,5 @@ X509_REQ_set0_signature                 ?	3_0_0	EXIST::FUNCTION:
 OSSL_PARAM_modified                     ?	3_0_0	EXIST::FUNCTION:
 OSSL_PARAM_set_unmodified               ?	3_0_0	EXIST::FUNCTION:
+EVP_PKEY_new_raw_private_key_with_libctx ?	3_0_0	EXIST::FUNCTION:
+EVP_PKEY_new_raw_public_key_with_libctx ?	3_0_0	EXIST::FUNCTION:
```
Please document them

mattcaswell replied:
Yes, of course. Done.

mattcaswell commented:
> I had no time yet to review, but I wanted to start asking if we could also add a test for THESE new APIs so that we can catch further problems also with other key types?

Of course. We already had some tests for the non-libctx versions so I extended them to test both variants.

mattcaswell commented:
All feedback responded to and updates applied. I have rebased to resolve some conflicts with master and added some commits to deal with the test failures.

This is now out of WIP. Please take another look.

slontis commented:
There are at least 3 different travis errors.

slontis commented on crypto/ec/ecx_backend.c:
```diff
@@ -18,6 +20,31 @@
+        X25519_public_from_private(key->pubkey, key->privkey);
+        break;
+    case ECX_KEY_TYPE_ED25519:
+        if (!ED25519_public_from_private(key->libctx, key->pubkey, key->privkey)) {
```
Should then functions be passing a property_query string also?

mattcaswell replied:
This is definitely a non-trivial task. The problem is that this code is called provider side. If we're going to use the property query here then we'd have to pass it through as an OSSL_PARAM, and presumably have an `ecx_set_params` implementation in order to be able to associate that query with the key in some way. I suspect this problem is bigger than just ECX (we're quite inconsistent about where we pass both a libctx and propq and where we just pass a libctx) - so I suspect we're going to need to address that as a separate issue. Solving this in this case is straying quite far from my original scope for this PR - so I'd rather not try and deal with this here.

paulidale replied:
We pass property queries without the libctx in places.  KDF definitely but I think MAC too.

slontis commented on crypto/ec/ecx_key.c:
```diff
@@ -10,13 +10,14 @@
 #include "crypto/ecx.h"
 
-ECX_KEY *ecx_key_new(ECX_KEY_TYPE type, int haspubkey)
+ECX_KEY *ecx_key_new(OPENSSL_CTX *libctx, ECX_KEY_TYPE type, int haspubkey)
```
prop_query also??

mattcaswell replied:
As above

slontis commented on crypto/evp/p_lib.c:
```diff
@@ -318,78 +319,210 @@ int EVP_PKEY_cmp(const EVP_PKEY *a, const EVP_PKEY *b)
+        if (tmpe == NULL)
+            ameth = NULL;
+
+        ENGINE_finish(tmpe);
```
Engine API's need to be ifdef

mattcaswell replied:
Fixed

slontis commented on crypto/evp/p_lib.c:
```diff
@@ -318,78 +319,210 @@ int EVP_PKEY_cmp(const EVP_PKEY *a, const EVP_PKEY *b)
+            OSSL_PARAM params[] = { OSSL_PARAM_END, OSSL_PARAM_END };
+
+            ERR_clear_last_mark();
+            params[0] = OSSL_PARAM_construct_octet_string(
```
Will this actually work as an octet string???
ecx is the only one that uses octet string for both priv and pub.
ec uses BN for priv.
dsa and dh use BN for both.

slontis replied:
Ahh I see the docs say only weird types are used :)

mattcaswell replied:
Yes...this is for raw keys which are always just a string of bytes.

slontis commented on doc/man3/EVP_PKEY_new.pod:
```diff
@@ -46,16 +58,34 @@ EVP_PKEY_up_ref() increments the reference count of B<key>.
-will be automatically derived from the given private key data (if appropriate
-for the algorithm type).
+EVP_PKEY_new_raw_private_key_with_libctx() allocates a new B<EVP_PKEY>. Unless
+an engine should be used for the key type a provider for the key is found using
```
type,

mattcaswell replied:
Fixed

slontis commented on crypto/evp/p_lib.c:
```diff
@@ -318,78 +319,210 @@ int EVP_PKEY_cmp(const EVP_PKEY *a, const EVP_PKEY *b)
+                                 ENGINE *e,
+                                 const unsigned char *key,
+                                 size_t len,
+                                 int priv)
```
maybe rename this priv param to priv_enabled. If looks like a private key pointer otherwise.

mattcaswell replied:
I renamed it too `key_is_priv`

slontis commented on doc/man3/EVP_PKEY_new.pod:
```diff
@@ -46,16 +58,34 @@ EVP_PKEY_up_ref() increments the reference count of B<key>.
+
+EVP_PKEY_new_raw_private_key() does the same as
+EVP_PKEY_new_raw_private_key_with_libctx() except that the default library
+context is and default property query is used instead. If B<e> is non-NULL then
```
the default library context and default property query string are used.

mattcaswell replied:
Fixed

mattcaswell commented:
Fixups pushed addressing the feedback above. Hopefully should also fix the Travis issues.

openssl-machine commented:
This pull request is ready to merge

mattcaswell commented:
Unfortunately there was a conflict with master in ecx_meth.c - updated and force pushed.

@slontis please can you reconfirm?

slontis commented:
reconfirmed

mattcaswell commented:
Pushed. Thanks!
